### PR TITLE
Fix URI decoding for auth_digest

### PR DESCRIPTION
**Pull request by teoric**
decoding URIs in auth header breaks e.g. if it contains slashes

* **What kind of change does this PR introduce?**
  - [x] bug fix

* **What is the related issue number (starting with `#`)**

#1716

* **What is the current behavior?** (You can also link to an open issue here)

Authentication loop, see #1716

* **What is the new behavior (if this is a feature change)?**

Authentication, as before 14.2

* **Other information**:

* **Checklist**:

  - [x] I think the code is trivial

### CODE REVIEW

**Comment by jaraco** [comment]
I did some digging into the changes for 14.2, but according to the PR and the changelog, that would affect basic auth, not digest auth. How is that change implicated?

**Comment by teoric** [comment]
Please consider dda336459308b7d6f1a6c11cf7129e8779a0fae2 (_Add support for UTF-8 in **auth_digest** tool_), which introduced the change I am talking about.  I suppose the Changelog is just not up complete.  In the PR #1683, dda336459308b7d6f1a6c11cf7129e8779a0fae2 and three other changes to auth_digest [are listed](https://github.com/cherrypy/cherrypy/pull/1683#commits-pushed-5bc6fe5) among the changes on 22nd April.

**Comment by jaraco** [comment]
Thanks for the clarification. Now that I understand it a _bit_ better, I'm pretty sure this change isn't the right one. I'm not sure what the right change is, though. Let's continue the discussion in the ticket.

**Comment by webknjaz** [comment]
+1 I also felt like this is a weird place for trying to fix such issue.

**Comment by teoric** [comment]
Well, if it fixes a regression, would it not be sensible to accept a hacky fix first and search for the one true way later?

**Comment by webknjaz** [comment]
@teoric,

I'm really grateful that you spend efforts trying to figure this out and help us. We are short on maintainers and are really lucky to have such people like you contributing to the project and making it better!

This change, however, doesn't fix a regression, it monkey-patches what's related/unique to *your* concrete application, not CherryPy. And it patches the wrong place meaning that something in someone else's workflow gets broken. It's dangerous/harmful to throw in patches like this (even with no tests covering this case), this will likely cause another regression, which would be much harder to identify and eliminte.

*As maintainers,* @jaraco and I are responsible for the changes we accept and we do our best to avoid mistakes, which might be hard to fix in future. We have to weight every decision we make to ensure we don't introduce sudden regressions, don't break pre-decided design considerations, keep the code base in a good shape.

I hope that this issue gets fixed soon and as a quick workaround for your app I may suggest you to copy-paste this tool into your app, do whatever temporary patches necessary and use it instead of a built-in tool and then switch back to native tool once the fix is released.

Have a nice day and again thank you for your contributions!

**Comment by webknjaz** [comment]
I've added changes, which should completely eliminate the issue and take into account the format in which browsers send unicode digest auth

**Comment by teoric** [comment]
Thank you very much for unmonkeying and fixing this!

**Comment by webknjaz** [comment]
Could you please check this across browsers, at least FireFox, because I don't have one on the current machine.
Also, I'm going to wait until @jaraco reviews this again to double-check changes.

**Comment by teoric** [comment]
Works for me on Firefox and Chrome; thank you!

**Comment by webknjaz** [comment]
Great! Let's wait till @jaraco validates it as well then :)

File: cherrypy/lib/auth_digest.py
Comment (jaraco) [inline cherrypy/lib/auth_digest.py:149]: This code is clearly not optimal or intuitive... and int introduces two additional deprecated calls. I want to clean this up.

**Review by jaraco — APPROVED** [review]
Although I'm not happy with the decode routine, since this addresses the issue, I'm happy to accept it to address the regression.

File: cherrypy/lib/auth_digest.py
Comment (webknjaz) [reply cherrypy/lib/auth_digest.py:149]: Yeah, I tried that. But it's a can of worms, there's too many moving parts. Let's do refactoring separately from the bug fix, please.

**Pull request merged**
